src/storage/txn/flow_controller.rs

```diff
@@ -0,0 +1,1055 @@
+                > SMOOTHER_STALE_RECORD_THRESHOLD
+            {
+                let v = self.records.pop_front().unwrap().0;
+                self.total -= v;
+            } else {
+                break;
+            }
+        }
+    }
+
+    pub fn get_recent(&self) -> u64 {
+        if self.records.is_empty() {
+            return 0;
+        }
+        self.records.back().unwrap().0
+    }
+
+    pub fn get_avg(&self) -> f64 {
+        if self.records.is_empty() {
+            return 0.0;
+        }
+        self.total as f64 / self.records.len() as f64
+    }
+
+    pub fn get_max(&self) -> u64 {
+        if self.records.is_empty() {
+            return 0;
+        }
+        self.records.iter().max_by_key(|(k, _)| k).unwrap().0
+    }
+
+    pub fn get_percentile_90(&mut self) -> u64 {
+        if self.records.is_empty() {
+            return 0;
+        }
+        let mut v = self.records.make_contiguous().to_vec();
+        v.sort_by_key(|k| k.0);
+        v[((self.records.len() - 1) as f64 * 0.90) as usize].0
+    }
+
+    pub fn slope(&self) -> f64 {
+        if self.records.len() <= 1 {
+            return 0.0;
+        }
+
+        // calculate the average of left and right parts
+        let half = self.records.len() / 2;
+        let mut left = 0;
+        let mut right = 0;
+        for (i, r) in self.records.iter().enumerate() {
+            if i < half {
+                left += r.0;
+            } else if self.records.len() - i - 1 < half {
+                right += r.0;
+            }
+        }
+        // use the two averages with the time span of oldest and latest records
+        // to get a slope
+        let elapsed = duration_to_sec(
+            self.records
+                .back()
+                .unwrap()
+                .1
+                .duration_since(self.records.front().unwrap().1),
+        );
+        (right - left) as f64 / half as f64 / (elapsed / 2.0)
+    }
+
+    pub fn trend(&self) -> Trend {
+        if self.records.len() <= 1 {
+            return Trend::NoTrend;
+        }
+
+        // calculate the average of left and right parts
+        let half = self.records.len() / 2;
+        let mut left = 0;
+        let mut right = 0;
+        for (i, r) in self.records.iter().enumerate() {
+            if i < half {
+                left += r.0;
+            } else if self.records.len() - i - 1 < half {
+                right += r.0;
+            }
+        }
+
+        // decide if there is a trend by the two averages
+        // adding 2 here is to give a tolerance
+        if right > left + 2 {
+            Trend::Increasing
+        } else if right < left - 2 {
+            Trend::Decreasing
+        } else {
+            Trend::NoTrend
+        }
+    }
+}
+
+// CFFlowChecker records some statistics and states related to one CF.
+// These statistics falls into five categories:
+//   * memtable
+//   * L0 files
+//   * flush flow
+//   * L0 flow (compaction read flow of L0)
```

Comment:
These two names are hard to understand without reading the code. Maybe L0 consumption/production, L0 increase/decrease? I think we can hide RocksDB concept from algorithm as much as possible.